File: src/levels/L10-observer/chain.rb
class Order
  def initialize(first_handler) = @handler = first_handler

  # En una cadena, el primero que atiende se queda con el aviso: los demás nunca se enteran.
  # region: Order#complete!
  def complete!
    @status = :listo
    @handler.handle(self)
  end
  # endregion
end

File: src/levels/L10-observer/ts/chain.ts
interface Handler {
  handle(order: Order): void
}

class Order {
  private status = 'pendiente'

  constructor(private handler: Handler) {}

  // En una cadena, el primero que atiende se queda con el aviso: los demás nunca se enteran.
  // region: Order#complete!
  complete() {
    this.status = 'listo'
    this.handler.handle(this)
  }
  // endregion
}
